[quant][observer] Add l2 norm minimization

hx89 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#24022 [quant][observer] Add l2 norm minimization**

In histogram observer add an approximation for L2 error minimization for selecting min/max.
By selecting new min/max, we filter out outliers in input distribution.

This follows the implementation of NormMinimization::NonlinearQuantizationParamsSearch in caffe2/quantization/server/norm_minimization.cc

Differential Revision: [D16713239](https://our.internmc.facebook.com/intern/diff/D16713239/)

Commit:
[quant][observer] Add l2 norm minimization

In histogram observer add an approximation for L2 error minimization for selecting min/max.
By selecting new min/max, we filter out outliers in input distribution.

This follows the implementation of NormMinimization::NonlinearQuantizationParamsSearch in caffe2/quantization/server/norm_minimization.cc

Differential Revision: [D16713239](https://our.internmc.facebook.com/intern/diff/D16713239/)

Commit:
Update on "[quant][observer] Add l2 norm minimization"


In histogram observer add an approximation for L2 error minimization for selecting min/max.
By selecting new min/max, we filter out outliers in input distribution.

This follows the implementation of NormMinimization::NonlinearQuantizationParamsSearch in caffe2/quantization/server/norm_minimization.cc

Differential Revision: [D16713239](https://our.internmc.facebook.com/intern/diff/D16713239/)

Commit:
Update on "[quant][observer] Add l2 norm minimization"


In histogram observer add an approximation for L2 error minimization for selecting min/max.
By selecting new min/max, we filter out outliers in input distribution.

This follows the implementation of NormMinimization::NonlinearQuantizationParamsSearch in caffe2/quantization/server/norm_minimization.cc

Differential Revision: [D16713239](https://our.internmc.facebook.com/intern/diff/D16713239/)

Commit:
Update on "[quant][observer] Add l2 norm minimization"


In histogram observer add an approximation for L2 error minimization for selecting min/max.
By selecting new min/max, we filter out outliers in input distribution.

This follows the implementation of NormMinimization::NonlinearQuantizationParamsSearch in caffe2/quantization/server/norm_minimization.cc

Differential Revision: [D16713239](https://our.internmc.facebook.com/intern/diff/D16713239/)

Commit:
Update on "[quant][observer] Add l2 norm minimization"


In histogram observer add an approximation for L2 error minimization for selecting min/max.
By selecting new min/max, we filter out outliers in input distribution.

This follows the implementation of NormMinimization::NonlinearQuantizationParamsSearch in caffe2/quantization/server/norm_minimization.cc

Differential Revision: [D16713239](https://our.internmc.facebook.com/intern/diff/D16713239/)

Commit:
Update on "[quant][observer] Add l2 norm minimization"


In histogram observer add an approximation for L2 error minimization for selecting min/max.
By selecting new min/max, we filter out outliers in input distribution.

This follows the implementation of NormMinimization::NonlinearQuantizationParamsSearch in caffe2/quantization/server/norm_minimization.cc

Differential Revision: [D16713239](https://our.internmc.facebook.com/intern/diff/D16713239/)

Commit:
Update on "[quant][observer] Add l2 norm minimization"


In histogram observer add an approximation for L2 error minimization for selecting min/max.
By selecting new min/max, we filter out outliers in input distribution.

This follows the implementation of NormMinimization::NonlinearQuantizationParamsSearch in caffe2/quantization/server/norm_minimization.cc

Differential Revision: [D16713239](https://our.internmc.facebook.com/intern/diff/D16713239/)

Commit:
Update on "[quant][observer] Add l2 norm minimization"


In histogram observer add an approximation for L2 error minimization for selecting min/max.
By selecting new min/max, we filter out outliers in input distribution.

This follows the implementation of NormMinimization::NonlinearQuantizationParamsSearch in caffe2/quantization/server/norm_minimization.cc

Differential Revision: [D16713239](https://our.internmc.facebook.com/intern/diff/D16713239/)

Commit:
Update on "[quant][observer] Add l2 norm minimization"


In histogram observer add an approximation for L2 error minimization for selecting min/max.
By selecting new min/max, we filter out outliers in input distribution.

This follows the implementation of NormMinimization::NonlinearQuantizationParamsSearch in caffe2/quantization/server/norm_minimization.cc

Differential Revision: [D16713239](https://our.internmc.facebook.com/intern/diff/D16713239/)

Commit:
Update on "[quant][observer] Add l2 norm minimization"


In histogram observer add an approximation for L2 error minimization for selecting min/max.
By selecting new min/max, we filter out outliers in input distribution.

This follows the implementation of NormMinimization::NonlinearQuantizationParamsSearch in caffe2/quantization/server/norm_minimization.cc

Differential Revision: [D16713239](https://our.internmc.facebook.com/intern/diff/D16713239/)

Commit:
Update on "[quant][observer] Add l2 norm minimization"


In histogram observer add an approximation for L2 error minimization for selecting min/max.
By selecting new min/max, we filter out outliers in input distribution.

This follows the implementation of NormMinimization::NonlinearQuantizationParamsSearch in caffe2/quantization/server/norm_minimization.cc

Differential Revision: [D16713239](https://our.internmc.facebook.com/intern/diff/D16713239/)

Commit:
Update on "[quant][observer] Add l2 norm minimization"


In histogram observer add an approximation for L2 error minimization for selecting min/max.
By selecting new min/max, we filter out outliers in input distribution.

This follows the implementation of NormMinimization::NonlinearQuantizationParamsSearch in caffe2/quantization/server/norm_minimization.cc

Differential Revision: [D16713239](https://our.internmc.facebook.com/intern/diff/D16713239/)

Commit:
Update on "[quant][observer] Add l2 norm minimization"


In histogram observer add an approximation for L2 error minimization for selecting min/max.
By selecting new min/max, we filter out outliers in input distribution.

This follows the implementation of NormMinimization::NonlinearQuantizationParamsSearch in caffe2/quantization/server/norm_minimization.cc

Differential Revision: [D16713239](https://our.internmc.facebook.com/intern/diff/D16713239/)

lly-zero-one reviewed (changes requested):
Oh, saw your test in stacked PR.

Commit:
Update on "[quant][observer] Add l2 norm minimization"


In histogram observer add an approximation for L2 error minimization for selecting min/max.
By selecting new min/max, we filter out outliers in input distribution.

This follows the implementation of NormMinimization::NonlinearQuantizationParamsSearch in caffe2/quantization/server/norm_minimization.cc

Differential Revision: [D16713239](https://our.internmc.facebook.com/intern/diff/D16713239/)

Commit:
Update on "[quant][observer] Add l2 norm minimization"


In histogram observer add an approximation for L2 error minimization for selecting min/max.
By selecting new min/max, we filter out outliers in input distribution.

This follows the implementation of NormMinimization::NonlinearQuantizationParamsSearch in caffe2/quantization/server/norm_minimization.cc

Differential Revision: [D16713239](https://our.internmc.facebook.com/intern/diff/D16713239/)

Commit:
Update on "[quant][observer] Add l2 norm minimization"


In histogram observer add an approximation for L2 error minimization for selecting min/max.
By selecting new min/max, we filter out outliers in input distribution.

This follows the implementation of NormMinimization::NonlinearQuantizationParamsSearch in caffe2/quantization/server/norm_minimization.cc

Differential Revision: [D16713239](https://our.internmc.facebook.com/intern/diff/D16713239/)

Commit:
Update on "[quant][observer] Add l2 norm minimization"


In histogram observer add an approximation for L2 error minimization for selecting min/max.
By selecting new min/max, we filter out outliers in input distribution.

This follows the implementation of NormMinimization::NonlinearQuantizationParamsSearch in caffe2/quantization/server/norm_minimization.cc

Differential Revision: [D16713239](https://our.internmc.facebook.com/intern/diff/D16713239/)

Commit:
Update on "[quant][observer] Add l2 norm minimization"


In histogram observer add an approximation for L2 error minimization for selecting min/max.
By selecting new min/max, we filter out outliers in input distribution.

This follows the implementation of NormMinimization::NonlinearQuantizationParamsSearch in caffe2/quantization/server/norm_minimization.cc

Differential Revision: [D16713239](https://our.internmc.facebook.com/intern/diff/D16713239/)

Commit:
Update on "[quant][observer] Add l2 norm minimization"


In histogram observer add an approximation for L2 error minimization for selecting min/max.
By selecting new min/max, we filter out outliers in input distribution.

This follows the implementation of NormMinimization::NonlinearQuantizationParamsSearch in caffe2/quantization/server/norm_minimization.cc

Differential Revision: [D16713239](https://our.internmc.facebook.com/intern/diff/D16713239/)

Commit:
Update on "[quant][observer] Add l2 norm minimization"


In histogram observer add an approximation for L2 error minimization for selecting min/max.
By selecting new min/max, we filter out outliers in input distribution.

This follows the implementation of NormMinimization::NonlinearQuantizationParamsSearch in caffe2/quantization/server/norm_minimization.cc

Differential Revision: [D16713239](https://our.internmc.facebook.com/intern/diff/D16713239/)

raghuramank100 commented on torch/quantization/observer.py:
```diff
@@ -167,53 +168,283 @@ class HistogramObserver(ObserverBase):
+                self.min_val = min_val - 0.5 * range
+                self.max_val = max_val + 0.5 * range
+                self.histogram = torch.histc(
+                    x, self.bins, min=min_val - 0.5 * range, max=max_val + 0.5 * range
```
To ensure that autograd does not process this part please do:

```python
def forward(self, x):
 with torch.no_grad():
  < Body of function>
```

raghuramank100 commented on torch/quantization/observer.py:
```diff
@@ -167,53 +168,283 @@ class HistogramObserver(ObserverBase):
+            ) / 3
+        return density * norm
+
+    def _include_zero(self):
```
Please remove _include_zero method.

raghuramank100 commented on torch/quantization/observer.py:
```diff
@@ -167,53 +168,283 @@ class HistogramObserver(ObserverBase):
+        caffe2/quantization/server/norm_minimization.cc
+        """
+        assert self.histogram.size()[0] == self.bins, "bins mistmatch"
+        self._include_zero()
```
Ditto, we can ensure this in min/max calculation

raghuramank100 commented on torch/quantization/observer.py:
```diff
@@ -167,53 +168,283 @@ class HistogramObserver(ObserverBase):
-            self.histogram = torch.histc(
-                x, self.bins, min=min_val - 0.5 * range, max=max_val + 0.5 * range
-            )
+            if self.use_fixed_width is True:
```
Please ensure that min and max include zero prior to creating histograms.

hx89 replied:
Since eventually _calculate_qparams() will ensure min and max include zero, I think we don't need to do it before creating histograms.

raghuramank100 reviewed (changes requested):
Looks great, please address remaining comments

raghuramank100 commented on torch/quantization/observer.py:
```diff
@@ -167,53 +168,283 @@ class HistogramObserver(ObserverBase):
+
+    def calculate_qparams(self, norm_type="L2"):
+        if self.histogram is None:
+            raise Exception("must run observer before calling calculate_qparams!")
```
Replace with warning:
``` python
            warnings.warn(
                "must run observer before calling calculate_qparams.\
                                    Returning default scale and zero point "
            )
            return torch.tensor([1.0]), torch.tensor([0])
```

raghuramank100 reviewed (changes requested):
Please address suggested changes, almost ready to go

raghuramank100 commented on test/test_quantization.py:
```diff
@@ -831,31 +831,55 @@ def test_tensor_observer_scriptable(self, qdtype, qscheme):
+                else:
+                    ref_scale = 0.0313725
+                    ref_zero_point = -128 if qdtype is torch.qint8 else 0
+
```
Do you have a test to ensure that L2 observer is scriptable?

hx89 replied:
Can I make it scriptable in another PR? Since making it scriptable is not trivial and could add extra code that complicate the logic of this PR.

dzhulgakov replied:
I don't think the observer itself necessarily needs to be scriptable. We can totally call python objects back from script and I think the quantization process would happen in python environment anyway

Commit:
Update on "[quant][observer] Add l2 norm minimization"


In histogram observer add an approximation for L2 error minimization for selecting min/max.
By selecting new min/max, we filter out outliers in input distribution.

This follows the implementation of NormMinimization::NonlinearQuantizationParamsSearch in caffe2/quantization/server/norm_minimization.cc

Differential Revision: [D16713239](https://our.internmc.facebook.com/intern/diff/D16713239/)

hx89 commented:
Addressed comments.

hx89 commented on torch/quantization/observer.py:
```diff
@@ -137,7 +141,304 @@ def calculate_qparams(self):
+            if min_val is None or max_val is None or histogram is None:
+                min_val = torch.min(x)
+                max_val = torch.max(x)
+                if self.use_fixed_width is True:
```
fixed_width is using the fixed width for the histogram, so each new input will be fit into the histogram whose width is the 2x range of the first batch. It's a less accurate way than combining histogram for each new input. It could be a little confusing to users, maybe better remove this option.

Commit:
Update on "[quant][observer] Add l2 norm minimization"


In histogram observer add an approximation for L2 error minimization for selecting min/max.
By selecting new min/max, we filter out outliers in input distribution.

This follows the implementation of NormMinimization::NonlinearQuantizationParamsSearch in caffe2/quantization/server/norm_minimization.cc

Differential Revision: [D16713239](https://our.internmc.facebook.com/intern/diff/D16713239/)

hx89 commented:
Addressed comments.

dzhulgakov reviewed (approved):
Looks good in general, though I didn't look through each bit of the histogram merging

Commit:
Update on "[quant][observer] Add l2 norm minimization"


In histogram observer add an approximation for L2 error minimization for selecting min/max.
By selecting new min/max, we filter out outliers in input distribution.

This follows the implementation of NormMinimization::NonlinearQuantizationParamsSearch in caffe2/quantization/server/norm_minimization.cc

Differential Revision: [D16713239](https://our.internmc.facebook.com/intern/diff/D16713239/)

[ghstack-poisoned]

Commit:
Update on "[quant][observer] Add l2 norm minimization"


In histogram observer add an approximation for L2 error minimization for selecting min/max.
By selecting new min/max, we filter out outliers in input distribution.

This follows the implementation of NormMinimization::NonlinearQuantizationParamsSearch in caffe2/quantization/server/norm_minimization.cc

Differential Revision: [D16713239](https://our.internmc.facebook.com/intern/diff/D16713239/)

[ghstack-poisoned]

Commit:
Update on "[quant][observer] Add l2 norm minimization"


In histogram observer add an approximation for L2 error minimization for selecting min/max.
By selecting new min/max, we filter out outliers in input distribution.

This follows the implementation of NormMinimization::NonlinearQuantizationParamsSearch in caffe2/quantization/server/norm_minimization.cc

Differential Revision: [D16713239](https://our.internmc.facebook.com/intern/diff/D16713239/)

[ghstack-poisoned]

Commit:
Update on "[quant][observer] Add l2 norm minimization"


In histogram observer add an approximation for L2 error minimization for selecting min/max.
By selecting new min/max, we filter out outliers in input distribution.

This follows the implementation of NormMinimization::NonlinearQuantizationParamsSearch in caffe2/quantization/server/norm_minimization.cc

Differential Revision: [D16713239](https://our.internmc.facebook.com/intern/diff/D16713239/)

[ghstack-poisoned]

Commit:
Update on "[quant][observer] Add l2 norm minimization"


In histogram observer add an approximation for L2 error minimization for selecting min/max.
By selecting new min/max, we filter out outliers in input distribution.

This follows the implementation of NormMinimization::NonlinearQuantizationParamsSearch in caffe2/quantization/server/norm_minimization.cc

Differential Revision: [D16713239](https://our.internmc.facebook.com/intern/diff/D16713239/)

[ghstack-poisoned]

Commit:
Update on "[quant][observer] Add l2 norm minimization"


In histogram observer add an approximation for L2 error minimization for selecting min/max.
By selecting new min/max, we filter out outliers in input distribution.

This follows the implementation of NormMinimization::NonlinearQuantizationParamsSearch in caffe2/quantization/server/norm_minimization.cc

Differential Revision: [D16713239](https://our.internmc.facebook.com/intern/diff/D16713239/)

[ghstack-poisoned]
